test/model/src/core/util.cpp

```diff
@@ -283,6 +284,57 @@ shared_memory::~shared_memory()
     }
 }
 
+/*
+ * decode_utf8 --
+ *     Decode a UTF-8 string into one-byte code points. Throw an exception on error.
+ */
+std::string
+decode_utf8(const std::string &str)
+{
+    /* Initialize the conversion. */
+    iconv_t conv = iconv_open("WCHAR_T", "UTF-8");
+    if (conv == (iconv_t)-1)
+        throw std::runtime_error(
+          "Cannot initialize charset conversion: " + std::string(wiredtiger_strerror(errno)));
+    at_cleanup close_conv([conv] { (void)iconv_close(conv); });
+
+    /* Copy the input buffer, since the conversion library needs a mutable buffer. */
+    size_t src_size = str.size();
+    std::vector<char> src(str.begin(), str.end());
+
+    /* Allocate the output buffer. */
+    size_t dest_size = src_size + 4; /* Big enough because we're using 1-byte code points. */
+    std::vector<wchar_t> dest(dest_size, 0);
+
+    /* Now do the actual conversion. */
+    char *p_src = src.data();
+    char *p_dest = (char *)dest.data();
+    size_t src_bytes = src_size;
+    size_t dest_bytes = dest_size * sizeof(wchar_t);
+    size_t dest_bytes_start = dest_bytes;
+    size_t r = iconv(conv, &p_src, &src_bytes, &p_dest, &dest_bytes);
+    if (r == (size_t)-1)
+        throw std::runtime_error(
+          "Charset conversion failed: " + std::string(wiredtiger_strerror(errno)));
+    if (src_bytes != 0)
+        throw std::runtime_error(
+          "Charset conversion did not decode " + std::to_string(src_bytes) + " byte(s)");
+
+    /* Figure out how many code points were actually decoded. */
+    size_t decoded_size = (dest_bytes_start - dest_bytes) / sizeof(wchar_t);
+
+    /* Extract the byte-long code points into an array. */
+    std::vector<char> decoded(decoded_size, 0);
+    for (size_t i = 0; i < decoded_size; i++) {
+        if (dest[i] > 0xFF)
```

Comment:
I don't understand this logic, there are multibyte UTF-8 sequences that do not include 0xff, like UTF-8 2 byte characters: byte 1 = \xc0-\xdf, byte 2 = \x80-\xbf.

From what I can see, what does distinguish them is the high bit is set in both bytes.

Reply:
The `dest` array is the output of `iconv`'s conversion from UTF8 to `wchar_t`, so it already contains the decoded characters.

The use case for this function is to get a char array from `wt printlog -u`, which encodes non-printable bytes in JSON using `\uXXXX`. For example, if a key has value 1, WT packs it as one byte 0x81, which `printlog` writes to the JSON as `key: "\u0081"`. The JSON parser would interpret this as two bytes C2 81, which will become the `src` array in this function. `iconv` will then produce `dest` array of length 1 with 0x81, and this `for` loop then copies this value into a `char` array. The output of this function is later passed to `data_value::unpack`, which decodes this back into integer 1.

`printlog` only encodes individual bytes using the `\u` notation, so every decoded value that we get at this point will be between 0 and 0xFF. This `if` statement checks that we didn't get any unexpected value outside of this range (and I'm pretty sure that `wchar_t` is unsigned, so no need to check for < 0).

I hope that makes sense... let me know if you have any more questions. Would it make sense for me to include this example somewhere as a comment?

Reply:
Yes! I was completely missing `dest` is wchar_t. I appreciate you taking the time.

Reply:
I'm glad that this helps :)

I added the following big comment to the top of the function in [7c0fa41](https://github.com/wiredtiger/wiredtiger/pull/10551/commits/7c0fa4147bd840cf6672f1322c9c1c953f9b9bfa):
```
    /*
     * The main use case for this function is to recover the exact byte sequence specified in a JSON
     * document produced by "wt printlog -u", which uses the \u notation to encode non-printable
     * bytes. For example, WiredTiger packs key 1 as byte 81 (hex), which results in the following
     * line in the output of printlog:
     *
     *     "key": "\u0081"
     *
     * Our JSON parser encodes these Unicode characters using UTF-8, so reading this key results in
     * a two-byte string C2 81. This function decodes this string back to 81, which the caller can
     * then use in data_value::unpack to get the original integer 1.
     *
     * We do this in two steps:
     *   1. Convert the UTF-8 string to a WCHAR_T array.
     *   2. Convert the WCHAR_T array to a char array.
     */
```